api/task_manager.cc

```diff
@@ -0,0 +1,202 @@
+                            if (filter_tasks(task, params)) {
+                                tm::task_stats ts;
+                                ts.task_id = task->get_status().id.to_sstring();
+                                ts.state = task->get_status().state;
+                                res.push_back(ts);
+                            }
+                        }
+                    }
+                }
+                return res;
+            }, std::vector<tm::task_stats>(), concat<tm::task_stats>);
+        }).then([] (const std::vector<tm::task_stats>& res) {
+            return make_ready_future<json::json_return_type>(res);
+        });
+    });
+
+    tm::get_task_status.set(r, [&ctx] (std::unique_ptr<request> req) {
+        utils::UUID id = utils::UUID{req->param["task_id"]};
+        return ctx.tm.map_reduce0([id] (task_manager& tm) -> status {
+            status s;
+            s.shard = this_shard_id();
+            s.number = tm.get_running_tasks_count();
+            auto [module, task] = tm.retrieve_task(id);
+            if (task) {
+                s.task_status = task->get_status();
+                s.module = module;
+                // TODO Deexie progress should be gathered recursively
+            }
+            return s;
+        }, status{}, [id] (status l, status r) {
+            if (l.task_status.id == id) {
+                l.number += r.number;
+                return l;
+            } else {
+                r.number += l.number;
+                return r;
+            }
+        }).then([id] (const status& res) {
+            if (res.task_status.id != id) {
+                return make_exception_future<json::json_return_type>(httpd::bad_param_exception("Invalid task id"));
+            }
+            return make_ready_future<json::json_return_type>(make_status(res));
+        });
+    });
+
+    tm::abort_task.set(r, [&ctx] (std::unique_ptr<request> req) {
+        utils::UUID id = utils::UUID{req->param["task_id"]};
+        return ctx.tm.invoke_on_all([id] (task_manager& tm) {
+            auto task = tm.get_task(id);
+            if (task) {
+                task->abort();
+            }
+        }).then([]{
+             return make_ready_future<json::json_return_type>(json_void());
+        });
+    });
+
+    tm::wait_task.set(r, [&ctx] (std::unique_ptr<request> req) {
+        utils::UUID id = utils::UUID{req->param["task_id"]};
+        return ctx.tm.map_reduce0([id] (task_manager& tm) {
+            return tm.wait_task(id).then([&tm] (auto mt) {
+                auto module = mt.first;
+                auto task = mt.second;
+                status s;
+                s.shard = this_shard_id();
+                s.number = tm.get_running_tasks_count();
+                if (task) {
+                    s.task_status = task->get_status();
+                    s.module = module;
+                    // TODO Deexie progress should be gathered recursively
```

Comment:
The task needs a shared_promise the api (and others) can use to wait on the `pr.get_future().`
When the task is done it should call `pr.set_value()`, after setting the error if the task completed with an error.
Or if the task failed to complete due to an error, it should call `pr.set_exception(<exception_ptr>)`

When the task is successfully aborted, it should call `set_exception(seastar::abort_requested_exception())`